lib/tokenizer/preprocessor.js

```diff
@@ -26,15 +31,34 @@ var Preprocessor = module.exports = function () {
 
     this.pos = -1;
     this.lastGapPos = -1;
+    this.lastCharPos = -1;
+    this.droppedBufferSize = 0;
 
     this.gapStack = [];
 
     this.skipNextNewLine = false;
 
     this.lastChunkWritten = false;
     this.endOfChunkHit = false;
+    this.bufferWaterline = DEFAULT_BUFFER_WATERLINE;
 };
 
+Object.defineProperty(Preprocessor.prototype, 'sourcePos', {
+    get: function () {
+        return this.droppedBufferSize + this.pos;
+    }
+});
+
+Preprocessor.prototype.dropParsedChunk = function () {
+    if (this.pos > this.bufferWaterline) {
```

Comment:
I meant calling drops only when we meet hibernation token. Then, this will work naturally for streamed input, and for the regular `.parse` you can just feed input in chunks (basically what it does now, but on the higher level).